File: python/council-sdk/council/agents.py
"""Agents namespace — agent registration and lifecycle management."""

from __future__ import annotations

from typing import Any, AsyncIterator, Optional, Union

from council.transport import Transport
from council.types import Agent, AgentAction, AgentCapability, CostSummary


class AgentsNamespace:
    """Manage agents registered on the Council platform.

    Usage::

        agent = await client.agents.register(
            workspace_id="ws_abc",
            name="ResearchBot",
            model="claude-3-opus",
        )
    """

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    # ── CRUD ───────────────────────────────────────────────────────────────

    async def register(
        self,
        *,
        workspace_id: str,
        name: str,
        model: Optional[str] = None,
        provider: Optional[str] = None,
        personality: Optional[str] = None,
        autonomy_level: Optional[int] = None,
        capabilities: Optional[list] = None,
        configuration: Optional[dict] = None,
        metadata: Optional[dict] = None,
    ) -> Agent:
        """Register a new agent in a workspace."""
        body: dict[str, Any] = {
            "workspaceId": workspace_id,
            "name": name,
        }
        if model is not None:
            body["model"] = model
        if provider is not None:
            body["provider"] = provider
        if personality is not None:
            body["personality"] = personality
        if autonomy_level is not None:
            body["autonomyLevel"] = autonomy_level
        if capabilities is not None:
            body["capabilities"] = [
                c.value if isinstance(c, AgentCapability) else c for c in capabilities
            ]
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata

        resp = await self._t.post("/api/agents", json=body)
        return _parse_agent(resp.get("data", resp))

    async def get(self, agent_id: str) -> Agent:
        """Get an agent by ID."""
        resp = await self._t.get(f"/api/agents/{agent_id}")
        return _parse_agent(resp.get("data", resp))

    async def list(
        self,
        *,
        workspace_id: Optional[str] = None,
        status: Optional[str] = None,
        capability: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> AsyncIterator[Agent]:
        """List agents with optional filters. Yields Agent objects."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspaceId"] = workspace_id
        if status:
            params["status"] = status

        resp = await self._t.get("/api/agents", params=params)
        agents_data = resp.get("data", [])

        if not isinstance(agents_data, list):
            agents_data = [agents_data]

        count = 0
        for item in agents_data:
            agent = _parse_agent(item)

            # Client-side filter for capability (if API doesn't support it)
            if capability and capability not in agent.capabilities:
                continue

            yield agent
            count += 1
            if limit and count >= limit:
                break

    async def update(
        self,
        agent_id: str,
        *,
        name: Optional[str] = None,
        model: Optional[str] = None,
        personality: Optional[str] = None,
        autonomy_level: Optional[int] = None,
        capabilities: Optional[list] = None,
        configuration: Optional[dict] = None,
        metadata: Optional[dict] = None,
        status: Optional[str] = None,
    ) -> Agent:
        """Update an agent's configuration."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if model is not None:
            body["model"] = model
        if personality is not None:
            body["personality"] = personality
        if autonomy_level is not None:
            body["autonomyLevel"] = autonomy_level
        if capabilities is not None:
            body["capabilities"] = [
                c.value if isinstance(c, AgentCapability) else c for c in capabilities
            ]
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata
        if status is not None:
            body["status"] = status

        resp = await self._t.put(f"/api/agents/{agent_id}", json=body)
        return _parse_agent(resp.get("data", resp))

    async def delete(self, agent_id: str) -> None:
        """Permanently delete an agent."""
        await self._t.delete(f"/api/agents/{agent_id}")

    # ── Lifecycle ──────────────────────────────────────────────────────────

    async def suspend(self, agent_id: str, *, reason: Optional[str] = None) -> Agent:
        """Suspend an agent."""
        body: dict[str, Any] = {"status": "suspended"}
        if reason:
            body["configuration"] = {"suspend_reason": reason}
        resp = await self._t.put(f"/api/agents/{agent_id}", json=body)
        return _parse_agent(resp.get("data", resp))

    async def reactivate(self, agent_id: str) -> Agent:
        """Reactivate a suspended agent."""
        resp = await self._t.put(
            f"/api/agents/{agent_id}",
            json={"status": "active"},
        )
        return _parse_agent(resp.get("data", resp))

    # ── Execution ──────────────────────────────────────────────────────────

    async def execute(
        self,
        agent_id: str,
        *,
        action: str,
        context: dict[str, Any],
        canvas_id: Optional[str] = None,
        system_prompt: Optional[str] = None,
        temperature: Optional[float] = None,
        max_tokens: Optional[int] = None,
    ) -> dict[str, Any]:
        """Execute an action on an agent."""
        body: dict[str, Any] = {
            "action": action,
            "context": context,
        }
        if canvas_id:
            body["canvasId"] = canvas_id
        if system_prompt:
            body["systemPrompt"] = system_prompt
        if temperature is not None:
            body["temperature"] = temperature
        if max_tokens is not None:
            body["maxTokens"] = max_tokens

        resp = await self._t.post(f"/api/agents/{agent_id}/execute", json=body)
        return resp.get("data", resp)

    async def get_actions(
        self,
        agent_id: str,
        *,
        limit: int = 50,
    ) -> list[AgentAction]:
        """Get recent actions for an agent."""
        resp = await self._t.get(
            f"/api/agents/{agent_id}/actions",
            params={"limit": limit},
        )
        data = resp.get("data", [])
        if not isinstance(data, list):
            return []
        return [_parse_action(item) for item in data]

    async def get_cost(self, agent_id: str) -> CostSummary:
        """Get the cost summary for an agent."""
        resp = await self._t.get(f"/api/agents/{agent_id}/cost")
        data = resp.get("data", {})
        return CostSummary(
            total_cost=data.get("totalCost", data.get("total_cost", 0)),
            total_tokens=data.get("totalTokens", data.get("total_tokens", 0)),
            action_count=data.get("actionCount", data.get("action_count", 0)),
            breakdown=data.get("breakdown", {}),
        )


# ── Helpers ────────────────────────────────────────────────────────────────────


def _parse_agent(data: dict[str, Any]) -> Agent:
    """Parse a raw API response into an Agent model."""
    return Agent(
        id=data.get("id", ""),
        name=data.get("name", ""),
        status=data.get("status", "idle"),
        model=data.get("model"),
        provider=data.get("provider"),
        personality=data.get("personality"),
        autonomy_level=data.get("autonomyLevel", data.get("autonomy_level", 3)),
        capabilities=data.get("capabilities", []),
        configuration=data.get("configuration", {}),
        metadata=data.get("metadata", {}),
        workspace_id=data.get("workspaceId", data.get("workspace_id")),
        created_at=data.get("createdAt", data.get("created_at")),
        updated_at=data.get("updatedAt", data.get("updated_at")),
        api_key=data.get("apiKey", data.get("api_key")),
    )


def _parse_action(data: dict[str, Any]) -> AgentAction:
    return AgentAction(
        id=data.get("id", ""),
        agent_id=data.get("agentId", data.get("agent_id", "")),
        action=data.get("action", ""),
        context=data.get("context", {}),
        response=data.get("response"),
        tokens_used=data.get("tokensUsed", data.get("tokens_used", 0)),
        cost=data.get("cost", 0),
        model=data.get("model"),
        created_at=data.get("createdAt", data.get("created_at")),
    )

File: typescript/src/namespaces/agents.ts
import { Transport } from "../transport.js";
import type {
  Agent,
  AgentAction,
  CostSummary,
  CreateAgentOptions,
  ExecuteAgentOptions,
  UpdateAgentOptions,
} from "../types.js";

/**
 * Agent registration and lifecycle management.
 */
export class AgentsNamespace {
  constructor(private transport: Transport) {}

  /**
   * Register a new agent in a workspace.
   */
  async register(options: CreateAgentOptions): Promise<Agent> {
    const body: Record<string, unknown> = {
      workspaceId: options.workspaceId,
      name: options.name,
    };
    if (options.model !== undefined) body.model = options.model;
    if (options.provider !== undefined) body.provider = options.provider;
    if (options.personality !== undefined)
      body.personality = options.personality;
    if (options.autonomyLevel !== undefined)
      body.autonomyLevel = options.autonomyLevel;
    if (options.capabilities !== undefined)
      body.capabilities = options.capabilities;
    if (options.configuration !== undefined)
      body.configuration = options.configuration;
    if (options.metadata !== undefined) body.metadata = options.metadata;

    const resp = await this.transport.post("/api/agents", { json: body });
    return parseAgent((resp.data ?? resp) as Record<string, unknown>);
  }

  /**
   * Get an agent by ID.
   */
  async get(agentId: string): Promise<Agent> {
    const resp = await this.transport.get(`/api/agents/${agentId}`);
    return parseAgent((resp.data ?? resp) as Record<string, unknown>);
  }

  /**
   * List agents with optional filters. Returns an async iterable.
   */
  async *list(
    options: {
      workspaceId?: string;
      status?: string;
      capability?: string;
      limit?: number;
    } = {},
  ): AsyncIterableIterator<Agent> {
    const params: Record<string, unknown> = {};
    if (options.workspaceId) params.workspaceId = options.workspaceId;
    if (options.status) params.status = options.status;

    const resp = await this.transport.get("/api/agents", { params });
    const data = resp.data;
    const agents = Array.isArray(data) ? data : [data];

    let count = 0;
    for (const item of agents) {
      const agent = parseAgent(item as Record<string, unknown>);

      if (
        options.capability &&
        !agent.capabilities.includes(options.capability)
      ) {
        continue;
      }

      yield agent;
      count++;
      if (options.limit && count >= options.limit) break;
    }
  }

  /**
   * Update an agent's configuration.
   */
  async update(agentId: string, options: UpdateAgentOptions): Promise<Agent> {
    const body: Record<string, unknown> = {};
    if (options.name !== undefined) body.name = options.name;
    if (options.model !== undefined) body.model = options.model;
    if (options.personality !== undefined)
      body.personality = options.personality;
    if (options.autonomyLevel !== undefined)
      body.autonomyLevel = options.autonomyLevel;
    if (options.capabilities !== undefined)
      body.capabilities = options.capabilities;
    if (options.configuration !== undefined)
      body.configuration = options.configuration;
    if (options.metadata !== undefined) body.metadata = options.metadata;
    if (options.status !== undefined) body.status = options.status;

    const resp = await this.transport.put(`/api/agents/${agentId}`, {
      json: body,
    });
    return parseAgent((resp.data ?? resp) as Record<string, unknown>);
  }

  /**
   * Permanently delete an agent.
   */
  async delete(agentId: string): Promise<void> {
    await this.transport.delete(`/api/agents/${agentId}`);
  }

  /**
   * Suspend an agent.
   */
  async suspend(
    agentId: string,
    options: { reason?: string } = {},
  ): Promise<Agent> {
    const body: Record<string, unknown> = { status: "suspended" };
    if (options.reason) {
      body.configuration = { suspend_reason: options.reason };
    }
    const resp = await this.transport.put(`/api/agents/${agentId}`, {
      json: body,
    });
    return parseAgent((resp.data ?? resp) as Record<string, unknown>);
  }

  /**
   * Reactivate a suspended agent.
   */
  async reactivate(agentId: string): Promise<Agent> {
    const resp = await this.transport.put(`/api/agents/${agentId}`, {
      json: { status: "active" },
    });
    return parseAgent((resp.data ?? resp) as Record<string, unknown>);
  }

  /**
   * Execute an action on an agent.
   */
  async execute(
    agentId: string,
    options: ExecuteAgentOptions,
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {
      action: options.action,
      context: options.context,
    };
    if (options.canvasId) body.canvasId = options.canvasId;
    if (options.systemPrompt) body.systemPrompt = options.systemPrompt;
    if (options.temperature !== undefined)
      body.temperature = options.temperature;
    if (options.maxTokens !== undefined) body.maxTokens = options.maxTokens;

    const resp = await this.transport.post(`/api/agents/${agentId}/execute`, {
      json: body,
    });
    return (resp.data ?? resp) as Record<string, unknown>;
  }

  /**
   * Get recent actions for an agent.
   */
  async getActions(
    agentId: string,
    options: { limit?: number } = {},
  ): Promise<AgentAction[]> {
    const resp = await this.transport.get(`/api/agents/${agentId}/actions`, {
      params: { limit: options.limit ?? 50 },
    });
    const data = resp.data;
    if (!Array.isArray(data)) return [];
    return data.map((item) => parseAction(item as Record<string, unknown>));
  }

  /**
   * Get the cost summary for an agent.
   */
  async getCost(agentId: string): Promise<CostSummary> {
    const resp = await this.transport.get(`/api/agents/${agentId}/cost`);
    const data = (resp.data ?? resp) as Record<string, unknown>;
    return {
      totalCost: (data.totalCost ?? data.total_cost ?? 0) as number,
      totalTokens: (data.totalTokens ?? data.total_tokens ?? 0) as number,
      actionCount: (data.actionCount ?? data.action_count ?? 0) as number,
      breakdown: (data.breakdown ?? {}) as Record<string, unknown>,
    };
  }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function parseAgent(data: Record<string, unknown>): Agent {
  return {
    id: String(data.id ?? ""),
    name: String(data.name ?? ""),
    status: String(data.status ?? "idle"),
    model: data.model as string | undefined,
    provider: data.provider as string | undefined,
    personality: data.personality as string | undefined,
    autonomyLevel: (data.autonomyLevel ?? data.autonomy_level ?? 3) as number,
    capabilities: (data.capabilities ?? []) as string[],
    configuration: (data.configuration ?? {}) as Record<string, unknown>,
    metadata: (data.metadata ?? {}) as Record<string, unknown>,
    workspaceId: (data.workspaceId ?? data.workspace_id) as string | undefined,
    createdAt: data.createdAt ? new Date(data.createdAt as string) : undefined,
    updatedAt: data.updatedAt ? new Date(data.updatedAt as string) : undefined,
    apiKey: (data.apiKey ?? data.api_key) as string | undefined,
  };
}

function parseAction(data: Record<string, unknown>): AgentAction {
  return {
    id: String(data.id ?? ""),
    agentId: String(data.agentId ?? data.agent_id ?? ""),
    action: String(data.action ?? ""),
    context: (data.context ?? {}) as Record<string, unknown>,
    response: data.response as string | undefined,
    tokensUsed: (data.tokensUsed ?? data.tokens_used ?? 0) as number,
    cost: (data.cost ?? 0) as number,
    model: data.model as string | undefined,
    createdAt: data.createdAt ? new Date(data.createdAt as string) : undefined,
  };
}
